api/v1alpha1/httproute_types.go

```diff
@@ -93,7 +93,9 @@ const (
 
 // HeaderType constants.
 const (
-	HeaderTypeExact = "Exact"
+	HeaderTypeExact                = "Exact"
+	HeaderTypeRegularExpression    = "RegularExpression"
+	HeaderTypeImplementionSpecific = "ImplementationSpecific"
```

Comment:
We discussed this in office hours. There was some interest in the following:

1. A single regex type will not enable cross-compatibility with each implementation. Would it make sense to have multiple regex types instead, each denoting the type of regex implementation. This might look something like `PosixRegex` or `PcreRegex`. I think this idea came from @Miciah originally.
2. Instead of `ImplementationSpecific`, would it make sense to support custom types that were implementation specific. The reasoning here was that each implementation might want to support multiple approaches that weren't common enough to be part of the spec here. 

These are more high level ideas meant to start a discussion than suggestions for a change, interested in your thoughts here.

Reply:
Thanks for the comment.

I didn't think this through and adopted what we had adopted for `path` matching, which probably needs more thought as well.

> Would it make sense to have multiple regex types instead, each denoting the type of regex implementation. This might look something like PosixRegex or PcreRegex.

I agree. Else we are repeating the Ingress v1beta1 issue. Are there other regex implementations that we need to care about?

> would it make sense to support custom types that were implementation specific. The reasoning here was that each implementation might want to support multiple approaches that weren't common enough to be part of the spec here.

Are there any examples of this category?
I do like this approach. However, my current understanding goes as: ImplementationSpecific is a kitchen sink for all sorts of custom behavior, Overtime, as we see patterns emerging and we can abstract the commons into the `extended` API surface.

Reply:
@hbagdi @robscott please don't make the same mistake we can see in Ingress. Please check https://github.com/kubernetes/ingress-nginx/issues/555 and https://docs.google.com/spreadsheets/d/1-oS8CKD4MtdEMar7QMUibCb-TUJ5ZIglRl2cLvaQ5zs/edit#gid=0

Anything related to regular expression should be considered as non-portable.

Also, I am not sure how productive it is to spend time defining multiple regex types without founding edge cases that users can think of, like support negative lookahead.

Reply:
I agree with your point, @aledbf.
The question I now see is if there is a real need for regex-based header value or not. If it is not widespread, then we can simply get away with a `Exact` and `ImplementationSpecific` options.